app/controllers/users/nbp_wallet_controller.rb

```diff
@@ -0,0 +1,101 @@
+
+      @template_url = "nmshd://tr##{truncated_reference}"
+      @truncated_reference = truncated_reference
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      redirect_to new_user_registration_path, alert: t('common.errors.generic_try_later')
+    end
+
+    def qr_code
+      send_data RQRCode::QRCode.new("nmshd://tr##{params[:truncated_reference]}").as_png(border_modules: 0), type: 'image/png'
+    end
+
+    def relationship_status
+      relationship = @connector.pending_relationship_for_nbp_uid(@provider_uid)
+
+      if relationship.present?
+        render json: {status: 'ready'}
+      else
+        render json: {status: 'waiting'}
+      end
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      redirect_to nbp_wallet_connect_users_path, alert: t('common.errors.generic')
+    end
+
+    def finalize
+      @relationship = @connector.pending_relationship_for_nbp_uid(@provider_uid)
+      abort_and_refresh and return if @relationship.blank?
+
+      accept_and_create_user
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      abort_and_refresh
+    end
+
+    private
+
+    def accept_and_create_user # rubocop:disable Metrics/AbcSize
+      relationship_changes = @relationship[:changes]
+      abort_and_refresh if relationship_changes.size != 1
+
+      user_attributes = @connector.parse_userdata_in_rel_change(relationship_changes.first)
+
+      @user = User.new_from_omniauth(user_attributes)
+      @user.skip_confirmation_notification!
+
+      unless @user.save
+        return abort_and_refresh(t('common.errors.model_not_created', model: User.model_name.human,
+          errors: @user.errors.full_messages.join(', ')))
+      end
+
+      @user.identities = [UserIdentity.new(omniauth_provider: 'nbp', provider_uid: @provider_uid),
+                          UserIdentity.new(omniauth_provider: 'enmeshed', provider_uid: @relationship[:peer])]
+
+      if @connector.accept_relationship_change(@relationship[:id], relationship_changes.first[:id])
+        @user.send_confirmation_instructions
+        flash[:notice] = t('devise.registrations.signed_up_but_unconfirmed')
+        redirect_to home_index_path
+      else
+        abort_and_refresh
+      end
+    end
+
+    def abort_and_refresh(reason = t('common.errors.generic'))
+      Rails.logger.debug { "NbpWalletController calling abort_and_refresh() due to: #{reason}" }
+      @connector.reject_relationship(@relationship) if @relationship.present?
+      @user.hard_destroy if @user.present?
+      redirect_to nbp_wallet_connect_users_path, alert: reason
+    end
+
+    def set_provider_details
+      @provider_uid = session[:omniauth_provider_uid]
+      if @provider_uid.blank?
+        redirect_to new_user_registration_path, alert: t('common.errors.generic')
+      end
+    end
+
+    def set_connector_api
```

Comment:
I would further check that we have the `nbp` provider configured and enabled (pseudo-code, maybe redirect instead):

```rb
raise X unless resource_class.omniauth_providers.include? :nbp
raise X unless Settings.omniauth.nbp&.enabled && Settings.omniauth.nbp.enmeshed.present?
```

Reply:
This moved into the ConnectorApi but now I do check that the values are configured. :+1: